test/mocha/spidermonkey.js

```diff
@@ -112,4 +115,57 @@ describe("spidermonkey export/import sanity test", function() {
             assert.strictEqual(counter_strings, tests[i].strings, "String count mismatch for test " + tests[i].input);
         }
     });
+
+    it("should output and parse ES6 code correctly", function() {
+        var code = fs.readFileSync("test/input/spidermonkey/input.js", "utf-8");
+        var uglify_ast = uglify.parse(code);
+        var moz_ast = uglify_ast.to_mozilla_ast();
+        var from_moz_ast = uglify.AST_Node.from_mozilla_ast(moz_ast);
+        assert.strictEqual(
+            from_moz_ast.print_to_string(),
+            uglify_ast.print_to_string()
+        );
+    });
+
+    it("should be capable of importing from acorn", function() {
+        var code = fs.readFileSync("test/input/spidermonkey/input.js", "utf-8");
+        var uglify_ast = uglify.parse(code);
+        var moz_ast = acorn.parse(code, {sourceType: 'module', ecmaVersion: 9});
+        var from_moz_ast = uglify.AST_Node.from_mozilla_ast(moz_ast);
+        assert.strictEqual(
+            from_moz_ast.print_to_string(),
+            uglify_ast.print_to_string()
+        );
+    });
+
+    it("should produce an AST compatible with escodegen", function() {
+        var code = fs.readFileSync("test/input/spidermonkey/input.js", "utf-8");
+        var uglify_ast = uglify.parse(code);
+        var moz_ast = uglify_ast.to_mozilla_ast();
+        assert.strictEqual(
+            escodegen.generate(moz_ast, {
+                format: {
+                    indent: {
+                        style: "",
+                    },
+                    newline: "",
+                    space:  "",
+                    quotes: "double"
+                }
+            })
+                .replace(/;}/g, "}")
+                .replace(/var {/g, "var{")
+                .replace(/var \[/g, "var[")
+                .replace(/const {/g, "const{")
+                .replace(/const \[/g, "const[")
+                .replace(/get \"/g, "get\"")
+                .replace(/set \"/g, "set\"")
+                .replace(/\[object Object\].\[object Object\]/g, "new.target")  // escodegen issue
+                .replace(/\(await x\)/, "await x")
```

Comment:
This `escodegen` output comparison looks pretty brittle.

Rather than comparing against `escodegen` output directly, couldn't you just reparse it with `acorn.parse`, convert to uglify AST, then `print_to_string()` that, and then compare against the original `uglify_ast.print_to_string()`?

Reply:
I thought about that, but it may hide errors in one of the conversion processes. This test actually found errors that were hiding between the output and the input.

Reply:
Not really because it's doing a full round trip with escodegen and acorn in the middle. If it arrives at the same result once in uglify AST form it's the same by definition.

No matter, other than the `await x` substitution it looks pretty harmless.

Reply:
I see what you mean now. I'll get to that :)